### ISSUE TITLE
Publish a node-red-contrib-enocean version that make use of enocean-js

### ISSUE THREAD

**Issue by shuguet** [enocean-js/enocean-js#62]
Hello team EnOcean-JS,

I purchased EnOcean modules expecting to use them with Node-RED, only to find out the latest version of Node-Red and the latest version of node-red-contrib-enocean aren't compatible, as the Node-Red module isn't using the new enocean-js library.

I've seen some messages from last December saying this should be available "soon" (a week was mentioned in https://github.com/enocean-js/node-red-contrib-enocean/issues/12).

Is there any news on that? Is there a way to donate/contribute to the project to make this happen faster?

Best regards,
S.

**Comment by Holger-Will** [comment]
hi, thanks for letting me know that there is demand for this. i' m a bit short on time currently, as you might have guessed. If you can code, you are more than welcome to contribute. I'm not very familiar with node-red. i'm more into enocean and js. 
the first implementation was done by @nabbl . If you are willing to get your hands dirty with this, i can help you getting started...

**Comment by shuguet** [comment]
Thanks @Holger-Will , I can code, just not in JS, and like you I simply don't have the time... hence why I  am offering to "sponsor"/donate to make this happen :)

https://github.com/nabbl is there any chance you could update your node-red module to work with the new library from enocean-js?

**Comment by Holger-Will** [comment]
ok then... i will try to get something done by tomorrow.

**Comment by Holger-Will** [comment]
sorry i will not finish anything today... my plan for the current implementation is to write various modules to implement the enocean protocol. in the second step there will be an [enocean over ip](https://www.enocean-alliance.org/wp-content/uploads/2018/04/EnOcean_over_IP_Specification_v0.9.pdf) implementation.

i would prefer to implement the node-red node based on the enocean over ip protocol...

or what do you think? does it make sense to also have the node-red node directly connect to an enocean USB stick?

**Comment by shuguet** [comment]
Hello Will,

I think the EnOcean over IP is a nice addition in the future, but for now,
I have the EnOcean USB stick, a few different machines (RPi 3 B+, Intel
NUCs, etc.) and I'd like to make use of at least one of those devices with
Node-Red directly using the EnOcean USB stick.

Given the current deployment of EnOcean versus EnOcean over IP, starting
with "raw" EnOcean makes more sense. When EnOcean over IP gateways and
other devices using that start being used/available more widely, then
adding EnOcean over IP will make sense.

Best regards,

--
Sylvain Huguet
http://www.linkedin.com/in/sylvainhuguet

**Comment by Holger-Will** [comment]
thanks for your input. please let me know about some usecases. do you want to receive and send telegrams? what is more important sending or receiving? can you live with manual teach in (opposed to automatic teach in) ?

**Comment by shuguet** [comment]
Use cases are quite traditional, I have EnOcean micro modules on my
roller-shutters, I also have a SoftRemote, Door/Window detectors for
open/close, temperature sensors, and I have RGBW LED controller (Sunricher)
and AFRISO WaterSensor on order.

I'd like to be able to receive telegrams from those equipment, act on them
through NodeRed (send notifications, take decisions, etc.) and also send
telegrams back (close the roller-shutters, control the LEDs, etc.).

I'm not sure what manual teach-in versus automatic teach-in would mean,
what's the difference? (I've not be able to run the previous version of the
Node-Red module, so I have literally no reference point).
I chose EnOcean mostly for the "battery free" operation and the fact it can
work without a gateway, but I'm finding it has a very limited adoption so
far :/

--
Sylvain Huguet
http://www.linkedin.com/in/sylvainhuguet

**Comment by Holger-Will** [comment]
i have a basic version up and running now. it is probably still a few days from publication....

> I'm not sure what manual teach-in versus automatic teach-in would mean,
> what's the difference? 

every enocean sensor has an address (4 Bytes). it is allways send with every data telegram.
 
There are basically 3 different types of data telegrams:

* 1BS (1 Byte of data)
* 4BS (4 Bytes of data)
* VLD (variable length)

how the data is encoded in these data bytes is described for each sensor type in its "enocean equipment profile " ([eep](https://www.enocean-alliance.org/wp-content/uploads/2018/02/EEP268_R3_Feb022018_public.pdf))

to make any sense of a telegram you have to know the sensors eep (in the form `rorg-func-type`, f.e. `a5-02-03`)

the **automatic**  teach in process basically works like this: 

* an actor is set into learn mode so it can receive teach in telegram.
* a sensor sends a teach in telegram.
* the teach in telegram tells the actor what of eep the sensor is.
* in the actor the sensor address and its eep is stored, so when a data telegram is received the actor known how to interpret the telegram.

manual teach in means you have to store the address and eep yourself (not coming from a teach in telegram)

**Comment by shuguet** [comment]
Node-Red provides "configuration nodes" (
https://nodered.org/docs/creating-nodes/config-nodes).

That could be a way to (manually, or later on using automatic teach in)
store a sensor address and its corresponding EEP and additional info.

Meaning as a short term solution, we could manually create a key/value
map/list for each sensor, with each key being a sensor address and the
corresponding value being the meta-data for that sensor (EEP,
"alias"/friendly name, etc.)
What do you think?

Then later on, if automatic teach in is added, that automatic teach-in way
would just "append" a key/value to that list automatically.

Best regards,

--
Sylvain Huguet
http://www.linkedin.com/in/sylvainhuguet

**Comment by Holger-Will** [comment]
i have added a branch and a PR for node-red-contrib-enocean so you can test the current implementation.
for now only receiving is implemented.
It currently works like this:

* there is a config node for connecting to the serialport
* the enocean-actor node can listen to a single sensor. if the properties `senderId` and `eep` are empty, the node will emit all teachIn telegram information
* the teachIn info holds the `senderId` and `eep`
* put this info into the fields of the node.
* from then on the node will emit decoded telegrams

so for every sensor you want to listen to you need a separate enocean-actor node.

let me know if you need any help getting this to work. it would be great to have your feedback on the implementation going forward.

**Comment by shuguet** [comment]
Thanks a lot Will!

I'll check this over this week-end.

Once I have Node-Red installed, what should I do to get this particular
branch/pull request installed?
I don't believe Node-Red will find/install that automagically from the
Palette installer?

Best regards,

--
Sylvain Huguet
http://www.linkedin.com/in/sylvainhuguet

**Comment by Holger-Will** [comment]
Windows, Linux or Mac?

> Once I have Node-Red installed, what should I do to get this particular
> branch/pull request installed?


```
git clone https://github.com/enocean-js/enocean-js.git
cd enocean-js
git checkout node-red-contrib-enocean
npm i
```

then go to the node_modules folder and delete the @enocean-js folder there... (... sorry i will clean that mess up later ;-) ) on linux this would be done like that:

```
cd node_modules
rm -rf ./@enocean-js

```

then get the complete path to packages/node-red-contrib-enocean

```
cd ~/.node-red
npm install /complete/path/to/packages/node-red-contrib-enocean
```
see [Testing your node in Node-Red](https://nodered.org/docs/creating-nodes/first-node)

to later update  the node to a newer version, go to the enocean-js folder and `git pull` and then restart node-red. (on windows you might have to reinstall inside the .node-red folder... idk)

P.S.: Will is my last name... ;-) you can call me Holger :-)

**Comment by Holger-Will** [comment]
coming along nicely. receiving should work well. Also sending generic 4BS telegrams as well as a specialized RPS Button is implemented now. @shuguet  Have you had the time to look at this yet?
If you have problems getting this to work, i might publish a beta today or on Friday that you can install as any other node.

**Comment by BitTourette** [comment]
Excellent work, guys. I'll try it as soon as it's released.

**Comment by Holger-Will** [comment]
just published a first beta. you can install it with

`npm i node-red-contrib-enocean@0.1.0-beta.1` inside your `.node-red` folder 

or wait a few minutes until it becomes available in the catalog. (i'm not sure how the node-red catalog handles beta versions though...)

usage is not documented yet. if you need need some help getting started, don't hesitate to ask.
I'd really like to hear your feedback on this! And i might need some help for the documentation and help section.

![image](https://user-images.githubusercontent.com/10807348/56074949-efccfe00-5dba-11e9-9a2a-bf47e0b62b18.png)

**Comment by Holger-Will** [comment]
... the Pallete Manager does not seem to like betas to much :-( so i published 0.1.0 now...

**Comment by Holger-Will** [comment]
... It's available in the Pallete Manager now \o/...

**Comment by wolfchild** [comment]
Great progress! I will do some tests over the weekend using some Eltako wall switches and a USB300 on a raspberry pi.

**Comment by Holger-Will** [comment]
see #88 for some hints on how to use the nodes...

**Comment by shuguet** [comment]
Thanks Holger, I'm having a look now (and sorry for the mistake with your first name!)

First problem I'm facing is how to identify the Sender Id?

I'm trying something very simple:
I have a SoftRemote (https://nodon.fr/en/nodon/enocean-soft-remote/) and I want to trigger a flow when a button is pressed.

The EEP is written on their website (F6-02-02), but how can I get the Sender Id of my remote?

Also, one side of the remote is currently linked to my Roller/Shutter (3 "in wall" modules, EEP D2-05-00 according to https://nodon.fr/en/nodon/the-enocean-roller-shutter-module/), I would like to get their Sender Id as well.

I ran the "simple.js" exemple from this repo on the RPi, here is what I get when I press 2 of the 4 buttons on the remote:
`root@raspberrypi:~/enocean-js/examples# node simple.js 
55.0007.07.01.7a.f6500032908e30.00ffffffff3600.04
55.0007.07.01.7a.f6000032908e20.00ffffffff3300.f6
55.0007.07.01.7a.f6700032908e30.00ffffffff3d00.6d
55.0007.07.01.7a.f6000032908e20.00ffffffff3c00.35`

I think the senderId is in there, but which part is it?

Best regards,
Sylvain.

**Comment by shuguet** [comment]
Ok, I think I found the SenderId using `eep.js` instead.
Seems to be `0032908e` in my case.

I still can't receive messages in NodeRed (while I see things coming in when I use either `simple.js` or `eep.js` manually on the same RPi machine.

I'm also on your Slack instance if you want, might be easier to debug.

Thanks a lot :)

**Comment by Holger-Will** [comment]
published a new version `0.2.0` with better error handling. this should help us finding out what's not working.

**Comment by Holger-Will** [comment]
published a version with automatic teach in `0.3.0`

**Comment by Holger-Will** [comment]
published 0.4.0 based on some very helpful input by @shuguet, thanks!

**Comment by Holger-Will** [comment]
new version 0.4.1 
* fix sending 4BS LRN Telegrams

**Comment by Holger-Will** [comment]
new version 0.4.2

* some more fixes for sending teach in
* show sensor info after automatic teach in
* more flexible encoding algorithm
* msg.meta holds even more meta data about the telegram

**Comment by Holger-Will** [comment]
new version v0.4.3

* moved parser and transformer stream to shared config node, to improve performance and use less memory

**Comment by Holger-Will** [comment]
new version v0.4.4

* in node help for all nodes
* some small cosmetic improvements

**Comment by Holger-Will** [comment]
the next update will break enocean-btn to make it more flexible

**Comment by Holger-Will** [comment]
new version v0.4.4

* new Button implementation
* code cleanup
* fix enocean-out (again) ;-) (should be working reliably now)

**Comment by Holger-Will** [comment]
i was on my local branch all the time, and totally forgot to publish the lib version that is compatible to the current node-red-contrib-enocean implementation. i did so now...

new version v0.4.6

* now works with enocen-js@0.0.1-beta.18 and has that as dependency.

**Comment by Holger-Will** [comment]
removed some console.log i forgot. i have now tested v0.4.7 purely installed from npm (should have done that way before...) and everything works well...

**Comment by Holger-Will** [comment]
new version 0.4.8

* include some help for `enocean-btn` in readme

**Comment by Holger-Will** [comment]
new version 0.4.9

* `enocean-btn` now encodes a 4-Rocker switch. so you can encode 128 switches = 512 Rockers = 1024 Buttons with one USB Stick.
* new generalized `enocean-out` behavior (no documentation yet). one enocean-ou for all outgoing telegrams.
* support for sending (simple) `UTE` teach-in telegrams

**Comment by Holger-Will** [comment]
now that a general enocean-out node is implemented, i'm thinking about creating spezial nodes for various sensor type. the enocean-btn could be connected to this output for example, but then we could have nodes for temperature sensors etc. as well...
Or even have dashboard buttons that connect to the output directly.

**Comment by Holger-Will** [comment]
prototype is in place... this will be huge changes and most likely all existing flows will break...

waiting with publication for a few days to get some more nodes done, and write some docs...

**Comment by Holger-Will** [comment]
v 0.4.13 

* add node name to `msg.meta` (#107 )

**Comment by Holger-Will** [comment]
`0.5.0-alpha.0` example flow

![image](https://user-images.githubusercontent.com/10807348/57296977-a11b3700-70ce-11e9-9a92-61a15b52cb12.png)

[{"id":"a6d29d13.3c30c","type":"tab","label":"Flow 4","disabled":false,"info":""},{"id":"a97f335.b0a6dd","type":"debug","z":"a6d29d13.3c30c","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","x":620,"y":400,"wires":[]},{"id":"cc35e6a6.347f58","type":"enocean-in","z":"a6d29d13.3c30c","serialport":"cfcf4fbc.4250e","name":"USB 300","x":80,"y":400,"wires":[["75dd71b9.5c90a"]],"info":"Light and Blind Control - Application Style 1"},{"id":"709e124a.54eb0c","type":"enocean-out","z":"a6d29d13.3c30c","serialport":"cfcf4fbc.4250e","name":"","x":600,"y":480,"wires":[]},{"id":"e543c942.f4dc98","type":"enocean-btn","z":"a6d29d13.3c30c","name":"","x":320,"y":480,"wires":[["709e124a.54eb0c","a97f335.b0a6dd"]]},{"id":"42c9d6ac.d7c2e8","type":"inject","z":"a6d29d13.3c30c","name":"ON","topic":"","payload":"{\"button\":0,\"channel\":1,\"event\":\"click\"}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":90,"y":460,"wires":[["e543c942.f4dc98"]]},{"id":"3450a788.324b98","type":"inject","z":"a6d29d13.3c30c","name":"OFF","topic":"","payload":"{\"button\":1,\"channel\":1,\"event\":\"click\"}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":90,"y":500,"wires":[["e543c942.f4dc98"]]},{"id":"75dd71b9.5c90a","type":"enocean-actor","z":"a6d29d13.3c30c","name":"","x":370,"y":340,"wires":[["a97f335.b0a6dd"]]},{"id":"b8da7324.ac706","type":"inject","z":"a6d29d13.3c30c","name":"LRN","topic":"","payload":"{\"type\":\"LRN\",\"duration\":3}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":90,"y":340,"wires":[["75dd71b9.5c90a"]]},{"id":"cfcf4fbc.4250e","type":"enocean-config-node","z":"","serialport":"/dev/ttyUSB0"}]

**Comment by spants** [comment]
I can't import the flow - I think there is a problem with it (formatting in copy/paste?)

**Comment by Holger-Will** [comment]
so there seems to be a bug in NRs flow exporter with inject-nodes that inject json.
I have fixed the flow by hand and made sure it imports... updated the above comment accordingly

**Comment by Holger-Will** [comment]
mmmhhh... strange, now that I want to file a bug report on NR, exporting seems to be fine and I can't reproduce the faulty export and the flow exports just fine... very very strange!?

**Comment by Holger-Will** [comment]
the new version `0.5.0alpha.1` should be quite stable now.
there are flow examples for incoming and outgoing data. see the [readme](https://github.com/enocean-js/enocean-js/tree/master/packages/node-red-contrib-enocean)
you can now:

* teach in and delete sensors automatically or manually using messages
* delete sensors from the ui
* change sensors name from the ui or using messages
* telegrams contain not only the sensors given name but also the actor name.

**Comment by Holger-Will** [comment]
with `0.5.0-alpha.2` i have now switched to a different message format (to fix bug #111 )
everything else should be the same

**Comment by Holger-Will** [comment]
now I'm thinking about implementing a new node: the `extractor` node.

I now have a node-red instance running on an old raspberryPi, using it as dashboard server for my home...
working with the nodes, I realize that I repeatedly use a function node to extract message values:

```
return {payload: msg.payload.HUM.value.toFixed(2),topic: msg.meta.name};
```
what do you think about a convenient node to extract values?

**Comment by spants** [comment]
maybe out of scope? You might not want to go too far with it - as long as people know how to use it :)

**Comment by Holger-Will** [comment]
@spants you are probably right. I imagined something like the split node with multiple outputs, but you can achieve the same thing using already existing nodes...

That brings me to another question, do you think we should get rid of the 4-rocker node as well? see #118

**Comment by Holger-Will** [comment]
oh, and I've just published the first beta. I believe the input and actor node is quite stable. i just want to think about the output node a bit more...

the rest is documentation and clean up, but as long as the API seems stable, we can go to 0.5.0 release...

**Comment by Holger-Will** [comment]
@spants consider the following function node:

```
var x = ["SP","TMP"]
function extract(name){
    return {
        payload: msg.payload[name].value.toFixed(2),
        topic: msg.meta.name
    }
}
return x.map(item=>extract(item));
```
now if you set the number of outputs to the length of the array, and specify in the array which fields you want extracted, you can reuse this to extarct all kinds of values... thats what i imagined the extarctor to work like...

**Comment by spants** [comment]
Looks good! I will have to try that

**Comment by Holger-Will** [comment]
now that the API is frozen, I have published the first stable version v0.5.0

there is still no useful documentation, even wrong documentation :-(
I will fix that and any bug you encounter with patches. 
Feature request will be implemented with minor releases... 

on our way to v1.0.0 \o/

**Comment by spants** [comment]
> @spants consider the following function node:
> 
> ```
> var x = ["SP","TMP"]
> function extract(name){
>     return {
>         payload: msg.payload[name].value.toFixed(2),
>         topic: msg.meta.name
>     }
> }
> return x.map(item=>extract(item));
> ```
> 
> now if you set the number of outputs to the length of the array, and specify in the array which fields you want extracted, you can reuse this to extract all kinds of values... thats what i imagined the extarctor to work like...

I'm not a programmer, so forgive silly questions. I think that this will only find the first item in X array as there is no loop and therefore does not build an array for output... So I should do parallel ones of these for each value? (CO2/HUM/TEMP in combined sensor)

**Comment by Holger-Will** [comment]
`x.map` creates an array from an array, mapping input items to output items. https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/map

**Comment by es101** [comment]
> `0.5.0-alpha.0` example flow
> 
> ![image](https://user-images.githubusercontent.com/10807348/57296977-a11b3700-70ce-11e9-9a92-61a15b52cb12.png)
> 
> ```
> [{"id":"a6d29d13.3c30c","type":"tab","label":"Flow 4","disabled":false,"info":""},{"id":"a97f335.b0a6dd","type":"debug","z":"a6d29d13.3c30c","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","x":820,"y":180,"wires":[]},{"id":"cc35e6a6.347f58","type":"enocean-in","z":"a6d29d13.3c30c","serialport":"cfcf4fbc.4250e","name":"USB 300","x":280,"y":180,"wires":[["75dd71b9.5c90a"]],"info":"Light and Blind Control - Application Style 1"},{"id":"709e124a.54eb0c","type":"enocean-out","z":"a6d29d13.3c30c","serialport":"cfcf4fbc.4250e","name":"","x":800,"y":260,"wires":[]},{"id":"e543c942.f4dc98","type":"enocean-btn","z":"a6d29d13.3c30c","name":"","x":520,"y":260,"wires":[["709e124a.54eb0c","a97f335.b0a6dd"]]},{"id":"42c9d6ac.d7c2e8","type":"inject","z":"a6d29d13.3c30c","name":"ON","topic":"","payload":"{\"button\":0,\"channel\":1,\"event\":\"click\"}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":290,"y":240,"wires":[["e543c942.f4dc98"]]},{"id":"3450a788.324b98","type":"inject","z":"a6d29d13.3c30c","name":"OFF","topic":"","payload":"{\"button\":1,\"channel\":1,\"event\":\"click\"}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":290,"y":280,"wires":[["e543c942.f4dc98"]]},{"id":"75dd71b9.5c90a","type":"enocean-actor","z":"a6d29d13.3c30c","name":"","x":570,"y":120,"wires":[["a97f335.b0a6dd"]]},{"id":"b8da7324.ac706","type":"inject","z":"a6d29d13.3c30c","name":"LRN","topic":"","payload":"{\"type\":\"LRN\",\"duration\":3}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":290,"y":120,"wires":[["75dd71b9.5c90a"]]},{"id":"cfcf4fbc.4250e","type":"enocean-config-node","z":"","serialport":"/dev/ttyUSB0"}]
> ```

Hello Holger,

I tried this example flow.  By clicking the button ON or OFF I see in Dolpin View :
ID                  RORG       DATA
FFB4B686     RPS            18
    "                   "                08

if i send 
{"button":1,"channel":6,"event":"click"}

object
payload: object
meta: object
eep: "f6-03-01"
channel: 6
type: "data"
status: 16
data: object
R1: 0
EB: 1

object
payload: object
meta: object
eep: "f6-03-01"
channel: 6
type: "data"
status: 0
data: object
R1: 0
EB: 0


But the actor programmed on channel 6 does not respond.


What is going wrong??

Regards

Erwin

**Comment by Holger-Will** [comment]
it might be connected to a bug I found earlier... 
I'm setting the LRN Bit for all data telegrams to 1 by default, so you don't have to all the time... 
unfortunately, RPS telegrams do not have an LRN Bit. That is not a problem for most actors. But it is for all actors that interpret them as `f6-02-03` and just check the whole Byte value and whether it's `0x10`, `0x30`, `0x50` or `0x70` (and rightly so...). 
Setting the LRN Bit on those telegrams just add the value 8 to the byte and so cannot be interpreted by these actors.

long story short: it should now work in `v0.5.4` which I just published with a patch. It should be in the palette manager now.

**Comment by spants** [comment]
With 0.5.2, the 4 way rocker (sending eep F6-03-01) is detected as a 2 way rocker F6-02-01 with my other enocean gateway..... it might be related to that?

**Comment by Holger-Will** [comment]
yes and no... all RPS telegrams can be interpreted as any other one. For example, setting the bits 3 and 4 on a byte `00110000` is `0x30`. Interpreted as `f6-02-01` Bit3 means button A0 and Bit4 means pressed. Interpreted as `f6-02-03` it also means A0 pressed. But actors that support only `f6-02-03` do not check the single bytes.

So the interpretation of all RPS (f6) telegrams is up to the actor.

**Comment by spants** [comment]
I understand (I deleted my msg before you replied so that I can test further).... 0.5.2 had a problem with the rocker on my other gateway. 0.5.4 seems to report correctly

**Comment by es101** [comment]
> it might be connected to a bug I found earlier...
> I'm setting the LRN Bit for all data telegrams to 1 by default, so you don't have to all the time...
> unfortunately, RPS telegrams do not have an LRN Bit. That is not a problem for most actors. But it is for all actors that interpret them as `f6-02-03` and just check the whole Byte value and whether it's `0x10`, `0x30`, `0x50` or `0x70` (and rightly so...).
> Setting the LRN Bit on those telegrams just add the value 8 to the byte and so cannot be interpreted by these actors.
> 
> long story short: it should now work in `v0.5.4` which I just published with a patch. It should be in the palette manager now.

Thank you for the quick reply, but that did not solve my problem.

I use Nodon SIN-2-2-0x actors with two channels eep: d2-01-12 and blindactors SIN-2RS-01 eep:d2-05-00.
Switching: 
Value = 0x00: switch off
Value between 0x01 and 0x64: switch off

With dolphin view I see ffb4b685 10/00 for button:0 and 30/00 for button:1 with status 10/00 (two telegrams each).

With enocean-switch I configured: baseid(ffb4b680), Channel: n, data: 0 or 1.
The actor is programmed to listen on ffb4b685 for channel 5.

Any idea what to to?

Regards

Erwin

**Comment by Holger-Will** [comment]
oh, you can not use the rocker node for  d2-* telegrams. it only sends f6-03-01... for everything else you have to build the telegrams yourself. i do have a SIN-2-2-01 here i can try something tomorrow...

**Comment by Holger-Will** [comment]
so it seems d2-01-12 is one of the more complex telegram types. I will probably have to change some things in the base implementation to get this to work... I'm building a little test setup. I'll get back to you when I have something useable

**Comment by Holger-Will** [comment]
@es101 I have now created a new issue for the nodon module where I will post my findings... please see #127 for more finfo

**Comment by Holger-Will** [comment]
published v0.6.0

* support for UTE teach in response for bidirectional teach in (experimental)
* fix eep d2-01-*
* fix generating 4BS teach in  telegrams

should be in the palette manager in 15 min.
